walletnode/services/userdataprocess/task.go

```diff
@@ -0,0 +1,398 @@
+	task.nodes = nodes
+
+	if task.request == nil {
+		// PROCESS TO RETRIEVE USERDATA FROM METADATA LAYER
+		if err := nodes.ReceiveUserdata(ctx, task.userpastelid); err != nil {
+			return err
+		} else {
+			// Post on result channel
+			node := nodes[0]
+			if node.ResultGet != nil {
+				task.resultChanGet <- node.ResultGet
+			} else {
+				return errors.Errorf("failed to receive userdata")
+			}
+
+			log.WithContext(ctx).Debug("Finished retrieve userdata")
+		}
+	} else {
+		// PROCESS TO SET/UPDATE USERDATA TO METADATA LAYER
+		// Get the previous block hash
+		// Get block num
+
+		// TODO: Unblock this part after pastelClient support it
+		/*
+			blockNum, err := task.pastelClient.GetBlockCount(ctx)
+			if err != nil {
+				return nil, errors.Errorf("failed to get block num: %w", err)
+			}
+
+			// Get block hash string
+			blockInfo, err := task.pastelClient.GetBlockVerbose1(ctx, blockNum)
+			if err != nil {
+				return nil, errors.Errorf("failed to get block info with given block num %d: %w", blockNum, err)
+			}*/
+
+		// Decode hash string to byte
+		task.request.PreviousBlockHash = "" /*blockInfo.Hash*/
+		if err != nil {
+			return errors.Errorf("failed to convert hash string %s to bytes: %w", "" /*blockInfo.Hash*/, err)
+		}
+
+		// Get the value of task.request.ArtistPastelIDPassphrase for sign data, then empty it in the request to make sure it not sent to supernodes
+		passphrase := task.request.ArtistPastelIDPassphrase
+		task.request.ArtistPastelIDPassphrase = ""
+		// Marshal task.request to byte array for signing
+		js, err := json.Marshal(task.request)
+		if err != nil {
+			return errors.Errorf("failed to encode request %w", err)
+		}
+
+		// Hash the request
+		hashvalue, err := userdata.Sha3256hash(js)
+		if err != nil {
+			return errors.Errorf("failed to hash request %w", err)
+		}
+
+		// Sign request with Wallet Node's pastelID and passphrase
+		signature, err := task.pastelClient.Sign(ctx, hashvalue, task.request.ArtistPastelID, passphrase)
+		if err != nil {
+			return errors.Errorf("failed to sign ticket %w", err)
+		}
+
+		userdata := &userdata.UserdataProcessRequestSigned{
+			Userdata:     task.request,
+			UserdataHash: hex.EncodeToString(hashvalue),
+			Signature:    hex.EncodeToString(signature),
+		}
+
+		// Send userdata to supernodes for storing in MDL's rqlite db.
+		if err := nodes.SendUserdata(ctx, userdata); err != nil {
+
+			return err
+		} else {
+			res, err := task.AggregateResult(ctx, nodes)
+			if err != nil {
+				return err
+			}
+			// Post on result channel
+			task.resultChan <- &res
+			log.WithContext(ctx).WithField("userdata_result", res).Debug("Posted userdata result")
+		}
+	}
+
+	// close the connections
+	for i := range nodes {
+		if err := nodes[i].Connection.Close(); err != nil {
+			return errors.Errorf("failed to close connection to node %s %w", nodes[i].PastelID(), err)
+		}
+	}
+
+	// Wait for all connections to disconnect.
+	return groupConnClose.Wait()
```

Comment:
better if we should check & wrap the error

Reply:
Actually no need to wait() here since the groupConnClose above already wait for it. Will return nil instead to match with art register flow.